src/main/kotlin/org/jitsi/jicofo/conference/source/ValidatingConferenceSourceMap.kt

```diff
@@ -18,67 +18,302 @@
+     * are silently ignored.
+     *
+     * If the attempt is unsuccessful, a [ValidationFailedException] is thrown and the state of the map remains
+     * unchanged.
+     *
+     * @throws ValidationFailedException if the attempt to add sources is unsuccessful. The state of the map remains
+     * unchanged.
+     * @return The sources that have been accepted and added.
+     */
     @Throws(ValidationFailedException::class)
     fun tryToAdd(owner: Jid?, sourcesToAdd: EndpointSourceSet): ConferenceSourceMap {
+        val existingSourceSet = this[owner] ?: EndpointSourceSet()
 
-        val (allSources, allGroups) = this.toMediaSourceMap()
-        val validator = SSRCValidator(
-            owner?.resourceOrEmpty.toString(), // This is used for logging only...
-            allSources,
-            allGroups,
-            ConferenceConfig.config.maxSsrcsPerUser,
-            this.logger
-        )
+        // Check for validity of the new SSRCs, and conflicts with other endpoints.
+        sourcesToAdd.sources.forEach { source ->
+            if (source.ssrc <= 0 || source.ssrc >= 0x1_0000_0000 ) {
+                throw InvalidSsrcException(source.ssrc)
+            }
+            if (ssrcToOwnerMap.containsKey(source.ssrc)) {
+                // Adding the same source for the same endpoint is also invalid.
+                throw SsrcAlreadyUsedException(source.ssrc)
+            }
+            source.msid?.let { msid ->
+                if (msidToOwnerMap.containsKey(msid) && msidToOwnerMap[msid] != owner) {
+                    throw MsidConflictException(msid)
+                }
+            }
+            if (existingSourceSet.sources.size + sourcesToAdd.sources.size > maxSsrcsPerUser) {
+                throw SsrcLimitExceededException(maxSsrcsPerUser)
+            }
+        }
+
+        // If we rejected any of the SSRCs we'd throw above. We accepted them all.
+        val resultingSources = existingSourceSet.sources + sourcesToAdd.sources
+        val resultingSsrcs = resultingSources.map { it.ssrc }.toSet()
 
-        val (sourcesToRemove2, groupsToRemove) = ConferenceSourceMap(owner, sourcesToAdd).toMediaSourceMap()
-        try {
-            val added = validator.tryAddSourcesAndGroups(sourcesToRemove2, groupsToRemove)
-            val addedConferenceSourceMap = fromMediaSourceMap(
-                added[0] as MediaSourceMap,
-                added[1] as MediaSourceGroupMap
-            )
-            add(addedConferenceSourceMap)
-            return addedConferenceSourceMap
-        } catch (e: InvalidSSRCsException) {
-            throw ValidationFailedException(e.message)
+        val acceptedGroups = mutableSetOf<SsrcGroup>()
+        sourcesToAdd.ssrcGroups.forEach {
+            // We just ignore empty groups and groups which have been signaled multiple times.
+            when {
+                it.ssrcs.isEmpty() -> logger.info("Empty group signaled, ignoring.")
```

Comment:
How noisy will these logs be? I'm wondering if they need to be `debug`.

Reply:
They are currently warnings :)